File: src/my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

using namespace std::placeholders;
 
class NumberCounterNode : public rclcpp::Node
{
public:
   NumberCounterNode() : Node("number_counter"), counter_(0) // constrctor initilazation is recommended
   {
    subscriber_ = this->create_subscription<example_interfaces::msg::Int64>(
        "number",10,
        std::bind(&NumberCounterNode::callbackNumber, this,_1));
    publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count",10);
    
    server_ = this->create_service<example_interfaces::srv::SetBool>("reset_counter",
               std::bind(&NumberCounterNode::callbackResetNumber,this,_1,_2));
    
   }
 
private:
   void callbackNumber(const example_interfaces::msg::Int64::SharedPtr msg) //means that the msg object is passed by value as a constant
   {
  
     counter_ += msg->data; //data accumaulation eg:If counter_ was 10 and msg->data is 5, after this line counter_ will be 15
                              //If the next message has msg->data as -3, counter_ will become 12

      RCLCPP_INFO(this->get_logger(), "Number of messages recived = %ld",counter_);

      auto msg2 = example_interfaces::msg::Int64();
      msg2.data = counter_;
      publisher_->publish(msg2);
      
        
   }
   void callbackResetNumber(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                            const example_interfaces::srv::SetBool::Response::SharedPtr response)
   {
      if (request->data == true)
      {
         counter_ = 0;
         response->message = "conter SET to zero";
         response->success = true;
      }
      else
      {
         response->message = "conter NOT SET to zero";
         response->success = false;
      }

   }

   rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
   rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
   std::int64_t counter_;
   rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
};
 
int main(int argc, char **argv)
{
   rclcpp::init(argc, argv);
   auto node = std::make_shared<NumberCounterNode>(); 
   rclcpp::spin(node);
   rclcpp::shutdown();
   return 0;
}

File: src/my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool
 
 
class NumberCounterNode(Node): 
    def __init__(self):
        super().__init__("number_counter") 
        self.subscriber_ = self.create_subscription(Int64,"number",self.callback_number,10)
        self.counter_ = 0
        self.get_logger().info("Counter Reciver started")
        self.publisher_ = self.create_publisher(Int64,"number_count",10)
        self.server_ = self.create_service(
            SetBool,"reset_counter",self.callback_reset_counter)


    def callback_number(self, msg:Int64):
        self.counter_ += 1
        self.get_logger().info(f"Number of messages recived = {self.counter_}")
        msg2 = Int64()
        msg2.data = self.counter_
        self.publisher_.publish(msg2)

    def callback_reset_counter(self, request: SetBool.Request, response: SetBool.Response):
        if request.data == True:
            self.counter_ = 0
            response.message = "number_counter SET to zero"
            response.success = True
        else:
            response.message = "number_counter NOT SET to zero"
            response.success = False
        return response
          
def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode() 
    rclpy.spin(node)
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()